Executing more than one Cogserver python Request causes a segfault

The second python Request submitted to the cogserver leads to a segmentation fault.

E.g.:
opencog> restapi.Start
opencog> pyshell.shell
Connection closed by foreign host.

The opposite order of the above leads to the same result, so the problem is not particular to a certain request. The same result seems to happen with any Request when a second request is entered.

[2015-04-06 22:30:34:493] [INFO] Creating python request restapi.Start
[2015-04-06 22:30:34:493] [INFO] Created python request restapi.Start
[2015-04-06 22:30:51:916] [INFO] Creating python request pyshell.shell
[2015-04-06 22:30:51:916] [ERROR] Caught signal 11 (Segmentation fault) on thread 140234834568960
	Stack Trace:
	2: ./opencog/server/cogserver (opencog::Logger::Base::~Base() +0x5e) [0x40815e]
	3: ./opencog/server/cogserver (sighand(int) +0xe9) [0x407799]
	4: /lib/x86_64-linux-gnu/libc.so.6 (x +0x36d40) [0x7f8b106aad40]
	5: /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0 (x +0x1b92c0) [0x7f8b0e2be2c0]
	6: /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0 (x +0x9) [0x7f8b0e24f329]
	7: /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0 (x +0x22) [0x7f8b0e2d4e02]
	8: /home/opencog/oc/build/opencog/cython/libPythonModule.so (opencog::PyRequest::PyRequest(opencog::CogServer&, std::string const&, std::string const&, opencog::RequestClassInfo*) +0x83) [0x7f8aff9b9c33]
	9: /home/opencog/oc/build/opencog/cython/libPythonModule.so (opencog::PythonRequestFactory::create(opencog::CogServer&) const +0x1db) [0x7f8aff9b619b]
	10: /home/opencog/oc/build/opencog/server/libserver.so (opencog::Registry<opencog::Request>::create(opencog::CogServer&, std::string const&) +0x106) [0x7f8b115f0526]
	11: /home/opencog/oc/build/opencog/server/libserver.so (opencog::ConsoleSocket::OnLine(std::string const&) +0x19b) [0x7f8b115fb8db]
	12: /home/opencog/oc/build/opencog/server/libserver.so (opencog::ServerSocket::handle_connection(opencog::ServerSocket*) +0x319) [0x7f8b115f8729]
	13: /usr/lib/x86_64-linux-gnu/libboost_thread.so.1.54.0 (x +0xba4a) [0x7f8b101b0a4a]
	14: /lib/x86_64-linux-gnu/libpthread.so.0 (x +0x8182) [0x7f8b0f744182]
	15: /lib/x86_64-linux-gnu/libc.so.6 (x +0x6d) [0x7f8b1076e47d]

--- a/opencog/cython/PyRequest.cc
+++ b/opencog/cython/PyRequest.cc
@@ -7,6 +7,8 @@
 
 using namespace opencog;
 
+static bool module_initialized = false;
+
 PyRequest::PyRequest(CogServer& cs,
                      const std::string& moduleName,
                      const std::string& className,
@@ -21,8 +23,12 @@ PyRequest::PyRequest(CogServer& cs,
     // NOTE: You need to call the import functions in each separate
     // shared library that accesses Cython defined api. If you don't
     // then you get a crash when you call an api function.
-    import_opencog__agent_finder();
-
+    if (!module_initialized)
+    {
+        import_opencog__agent_finder();
+        module_initialized = true;
+    }
+    
     // Call out to our helper module written in cython  NOTE: Cython api calls
     // defined "with gil" can be called without grabbing the GIL manually.
     _pyrequest = instantiate_request(moduleName, className, this);

--- a/opencog/cython/PyRequest.cc
+++ b/opencog/cython/PyRequest.cc
@@ -20,9 +20,9 @@ PyRequest::PyRequest(CogServer& cs,
     _moduleName = moduleName;
     _className = className;
 
-    // NOTE: You need to call the import functions in each separate
-    // shared library that accesses Cython defined api. If you don't
-    // then you get a crash when you call an api function.
+    // NOTE: You need to call the import functions once and only once
+    // in each separate shared library that accesses Cython defined api.
+    // If you don't then you get a crash when you call an api function.
     if (!module_initialized)
     {
         import_opencog__agent_finder();

--- a/tests/cython/utilities/utilities.py
+++ b/tests/cython/utilities/utilities.py
@@ -23,7 +23,7 @@ executed = False
 
 def add_link(atom1, atom2):
     global executed
-    link = ATOMSPACE.add_link(types.ListLink, [atom1, atom2])
+    link = ListLink(atom1, atom2)
     executed = True
     return link
 
@@ -48,7 +48,7 @@ else:
     print "add_link - did NOT execute"
 
 executed = False
-execute_atom(   atomspace,
+execute_atom( atomspace,
     ExecutionOutputLink( 
         GroundedSchemaNode("py: add_link"),
         ListLink(

